CTRL+Enter=submit and Escape=cancel

rondevous opened the pull request:
**Hightlights:**
- `ctrl+enter` to submit translation when you're typing; it still works for applying popular suggestion
- `esc` to cancel translation
- `ctrl+a` or `i` for adding translation, when text input is not in focus, but is already expanded, will now focus to the textbox, instead of resetting what you had typed.

Issues with this pull have been fixed. Please review.

rondevous committed:
CTRL+Enter=submit and Escape=cancel

rondevous commented:
New shortcut keys:

- `ctrl+enter` will submit your translation
- `esc` will cancel what you typed!

rondevous committed:
stop losing mah 'typed in mantra' just cuz I clicked outside that input

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -70,16 +70,53 @@ function scrollItems (down) {
+ * Clicks the delete icon
+ */
+/*
+function deleteTranslation () {
```
Why is this function commented out? Do you plan to use it in the future?

rondevous replied:
Yeah, but this requires some code for selecting suggestions. I thought about CTRL+1,2,3... for selecting. From there, Apply, Edit and Delete can be used for each suggestion item. Not every user can apply/delete the suggestions, so that needs proper detection as well. The alternative would be CTRL+Up/Down to select suggestions.

Up/Down is already used for scrolling through search results.

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -70,16 +70,53 @@ function scrollItems (down) {
+  if (!input_wrapper_collapsed)   // don't toggle 'add' and 'cancel'
+    document.getElementsByClassName('key-add-suggestion-header').item(0).click() 
+  else
+    document.getElementsByClassName('input form-control tr-form-control key-add-suggestion-field').item(0).focus() //fixme: throws undefined error
```
Does this actually do anything if it throws undefined?

rondevous replied:
No, not really, idk why it throws undefined, but I feel like it needs fixing xD

jurf replied:
No, I was asking a different thing. :D If it throws undefined, does it have any effect? Wouldn’t it be the same the line?

rondevous replied:
It doesn't have any effect, afaik.

I made it focus to the form input when you press CTRL+A, if it wasn't already in focus.

cuz sometimes, when you are in the middle of typing something, and you get distracted, and then return back to the browser with a mouse click somewhere that's not in the input box, and then you press CTRL+A, that input will collapse, and whatever you had typed will go away.

So this is just a precaution for using a keyboard shortcut you weren't fully aware of what it did. You can always ESC from the form and CTRL+A, if you wanna edit the default or whatever.

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -70,16 +70,53 @@ function scrollItems (down) {
+ * Clicks the cancel button
+ */
+function cancelTranslation () {
+  // restrict when cancel button can be pressed
```
Nitpick: golden rule of commenting code – comment why, not what you are doing. It’s pretty obvious you are restricting it.

(And an even smaller nitpick – sometimes you start comments with small letters, sometimes with big; unnecessary inconsistency imho.)

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -70,16 +70,53 @@ function scrollItems (down) {
+ * Clicks the submit button
+ */
+function submitTranslation () {
+  // restrict when submit button can be pressed
```
Ditto.

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -70,16 +70,53 @@ function scrollItems (down) {
 function addTranslation () {
-  document.getElementsByClassName('key-add-suggestion-header').item(0).click()
+  // don't click if input box is expanded
+  var input_wrapper_collapsed = (document.getElementsByClassName('key-add-suggestion-wrap')[0].className == 'key-add-suggestion-wrap');
```
Could be a constant.

jurf reviewed (changes requested):
Looks good, thanks for the work! Just a few style nitpicks, otherwise nice.

Just to be sure, did you run it through the Standard JS linter? If not, please do, it’ll save me a format commit later on.

rondevous committed:
Standard JS style, refactor, improve docs

rondevous commented:
whoops, I'm kinda new to this

rondevous committed:
Add author's name, don't prevent default for unmatched keys

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -66,20 +67,78 @@ function scrollItems (down) {
+ * @returns False if NOT collapsed
+ */
+function inputWrapCollapsed () {
+  if (document.getElementsByClassName('key-add-suggestion-wrap')[0].className === 'key-add-suggestion-wrap') { // checking if it is NOT collapsed
```
You can return the bool directly (also, you don’t need to comment it twice, it’s already in the documentation above):
```js
  return !(document.getElementsByClassName('key-add-suggestion-wrap')[0].className === 'key-add-suggestion-wrap')

```

rondevous replied:
Should I rename the function to `inputTranslationOpen`, so it's easier to understand the rest of the script?

rondevous replied:
`inputWrapOpen` ? (won't be understood by those who haven't seen the html of the website)

rondevous replied:
the working of the function will be changed accordingly!

jurf replied:
Since it returns a boolean, it’s good practice to prefix it with ‘is’/‘has’ (so it’s immediately clear what it does) – `isTranslationInputOpen()` sounds most natural to me working with that rule.

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -66,20 +67,78 @@ function scrollItems (down) {
+  // Prevent possible glitch: Don't cancel if wrapper is collapsed, but focus is in a 'form-control' element
+  if (!inputWrapCollapsed) {
+    // Don't allow input-forms to keep the focus after cancel or else shortcuts won't work
+    document.getElementsByClassName('btn btn-default form-cancel-btn').item(0).focus()
```
Curious, why are you focusing before clicking?

rondevous replied:
should I focus after the click then?

rondevous replied:
The keyboard shortcuts that had been prevented inside the input-form, don't work again after "clicking" CANCEL

That's because when the cancel button was simply "clicked" (with the JS command), the input-wrapper collapses, but the _focus still remains inside the input-form._ So whatever keys you pressed after 'cancel' would be typed into the input form, and not trigger any of the other shortcuts.

If the 'click' came from a mouse press, then the shortcuts still work because the focus changed.

jurf reviewed (changes requested):
Sorry for the late review, got stuck up with school work. Good work!

jurf commented on telegram-translation-shortcuts.user.js:
```diff
@@ -156,10 +235,10 @@ function handleShortcut (e) {
     // Cycle bindings
     case 'l':
-      if (!e.ctrlKey) cycleBindings(true)
+      if (!e.ctrlKey) cycleBindings(true); else matchedKey = false
```
Hm, I’ll probably refactor this whole part later, it got a bit too messy with the if-else clauses.

jurf replied:
(Don’t worry about it.)

rondevous committed:
Simpler Restrictions--Refactor--Update README

- Rename inputWrapCollapsed() to `isinputTranslationOpen() and focus to the input when called
- Submit and Cancel are restricted inside handleShortcut() by checking the activeElement
- Improved internal docs
- Update README with new shortcuts and group link

rondevous commented on telegram-translation-shortcuts.user.js:
```diff
@@ -66,20 +67,76 @@ function scrollItems (down) {
+ * Clicks the cancel button
+ */
+function cancelTranslation () {
+  // Prevent visual glitch: Don't cancel if input-wrapper is already closed, but focus somehow remains in the 'form-control' element
```
I checked and found out that, there are at least 2 other elements with 'form-control' in their className. That includes 'search' and 'login'

rondevous commented on telegram-translation-shortcuts.user.js:
```diff
@@ -66,20 +67,76 @@ function scrollItems (down) {
+/**
+ * Clicks the submit button
+ */
+function submitTranslation () {
+  // Prevent visual glitch: Don't re-submit if input-wrapper is already closed, but focus is still inside the 'form-control' element
+  if (inputTranslationOpen()) {
```
There are at least 2 other elements with 'form-control' in their className; which includes 'search' and 'login'. Was thinking about a restriction that would prevent submit/cancel when typing in those areas

jurf replied:
Maybe work with the focus instead? https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/activeElement

rondevous replied:
```
function handleShortcut (e) {
  // Handle shortcuts inside input forms
  if (e.target.classList.contains('form-control')) {
```
is it better to use activeElement here instead of e.target?

jurf replied:
Looks good.

jurf commented:
LGTM. Thanks a lot for the work (and the patience!).

jurf commented:
Oh, one last, pleasant, duty. Would you mind updating the readme to reflect your changes? :slightly_smiling_face:

rondevous commented on telegram-translation-shortcuts.user.js:
```diff
@@ -2,6 +2,7 @@
 // @description  Adds useful keyboard shortcuts to the Telegram Translation Platform
+// @author       Juraj Fiala
 // @include      https://translations.telegram.org/*
 // @version      0.3.2
```
New version after merge?

jurf replied:
Yeah, otherwise people who already have it installed won’t get the update.

rondevous replied:
no, I meant to ask if I should change the version number just before merge or in a separate commit after merge?

jurf commented:
I sent you the invite. Merge whenever you feel ready. :-)

rondevous commented:
doing a rebase of 5 commits now. lets see

rondevous commented:
After merging, **if Escape to cancel** translation doesn't work, please change the version to **0.4.1**
